agent/lib/kontena/workers/service_pod_worker.rb

```diff
@@ -62,7 +68,42 @@ def on_container_event(topic, e)
       if attrs['io.kontena.service.id'] == @service_pod.service_id &&
           attrs['io.kontena.service.instance_number'].to_i == @service_pod.instance_number
```

Comment:
The event also includes the `attrs['io.kontena.container.deploy_rev']`. Ignore the event if that doesn't match `@service_pod.deploy_rev`?

Reply:
That makes sense.

Reply:
Implemented better event check.

Reply:
Hmm, no, the `io.kontena.container.deploy_rev` doesn't work. The `@service_pod.deploy_rev` will change on each deploy without the container getting recreated, so the event `deploy_rev` label can be something much older. It's not useful for detecting old events from before the apply:

```
I, [2017-09-04T13:21:23.760931 #1]  INFO -- Kontena::Workers::ServicePodWorker: starting restart-test-1
D, [2017-09-04T13:21:24.148467 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event start: container.deploy_rev=2017-09-04 12:56:59 UTC service_pod.deploy_rev=2017-09-04 13:21:21 UTC)
D, [2017-09-04T13:21:24.149617 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event restart: container.deploy_rev=2017-09-04 12:56:59 UTC service_pod.deploy_rev=2017-09-04 13:21:21 UTC)
D, [2017-09-04T13:21:29.989317 #1] DEBUG -- Kontena::Workers::ServicePodWorker: on_container_event die: container.deploy_rev=2017-09-04 12:56:59 UTC service_pod.deploy_rev=2017-09-04 13:21:21 UTC)
```

I think it needs the `@container_id` approach, set from `apply`. They are not the same thing.

Reply:
Testing with something like this seems to work?

```patch
--- a/agent/lib/kontena/workers/service_pod_worker.rb
+++ b/agent/lib/kontena/workers/service_pod_worker.rb
@@ -65,14 +65,10 @@ module Kontena::Workers
 
     # @param [String] topic
     # @param [Docker::Event] e
-    def on_container_event(topic, e)
-      attrs = e.actor.attributes
-      if attrs['io.kontena.service.id'] == @service_pod.service_id &&
-          attrs['io.kontena.container.type'] == 'service' &&
-          attrs['io.kontena.service.instance_number'].to_i == @service_pod.instance_number &&
-          attrs['io.kontena.container.deploy_rev'].to_s == @service_pod.deploy_rev
+    def on_container_event(topic, event)
+      if @container && event.id == @container.id
         @container_state_changed = true
-        handle_restart_on_die if e.status == 'die'.freeze
+        handle_restart_on_die if event.status == 'die'.freeze
       end
     end
 
@@ -119,7 +115,7 @@ module Kontena::Workers
       cancel_restart_timers
       exclusive {
         begin
-          ensure_desired_state
+          @container = ensure_desired_state
           # reset restart counter if instance stays up 10s
           @restart_counter_reset_timer = after(10) {
             info "#{@service_pod.name_for_humans} stayed up 10s, resetting restart backoff counter" if restarting?
@@ -140,15 +136,16 @@ module Kontena::Workers
       }
     end
 
+    # @return [Docker::Container, nil]
     def ensure_desired_state
       debug "state of #{service_pod.name}: #{service_pod.desired_state}"
       service_container = get_container(service_pod.service_id, service_pod.instance_number)
       if service_pod.running? && service_container.nil?
         info "creating #{service_pod.name}"
-        ensure_running
+        service_container = ensure_running
       elsif service_pod.running? && (service_container && service_container_outdated?(service_container))
         info "re-creating #{service_pod.name}"
-        ensure_running
+        service_container = ensure_running
       elsif service_container && service_pod.running? && !service_container.running?
         info "starting #{service_pod.name}"
         ensure_started
@@ -158,6 +155,7 @@ module Kontena::Workers
       elsif service_pod.terminated?
         info "terminating #{service_pod.name}"
         ensure_terminated if service_container
+        service_container = nil
       elsif service_pod.desired_state_unknown?
         info "desired state is unknown for #{service_pod.name}, not doing anything"
       elsif state_in_sync?(service_pod, service_container)
@@ -165,8 +163,11 @@ module Kontena::Workers
       else
         warn "unknown state #{service_pod.desired_state} for #{service_pod.name}"
       end
+
+      service_container
     end
 
+    # @return [Docker::Container]
     def ensure_running
       Kontena::ServicePods::Creator.new(service_pod).perform
     rescue => exc
```